File: Classes/Cloud/Dropbox/FTDropboxManager.m
//
//  FTDropboxManager.m
//  Noteshelf
//
//  Created by Amar Udupa on 20/3/14.
//
//

#import "FTDropboxManager.h"
#import "FTBackUpAccountInfo.h"

@import ObjectiveDropboxOfficial;

CG_EXTERN NSString *const FTDidCompleteDropBoxAuthetication;
CG_EXTERN NSString *const FTDidCancelDropBoxAuthetication;

@interface FTDropboxManager()

@property (copy)FTDropboxAuthenticateCallback authCallBack;
@property (strong) FTBackUpAccountInfo *accountInfo;
@property (strong) NSMutableArray *accInfoCallBackArray;

@end

@implementation FTDropboxManager
@synthesize authCallBack;

static FTDropboxManager *dropboxManager = nil;

+ (instancetype)sharedDropboxManager
{
    static dispatch_once_t onceToken;
    dispatch_once(&onceToken, ^{
        dropboxManager = [[FTDropboxManager alloc] init];
        dropboxManager.accInfoCallBackArray = [NSMutableArray array];
        [[NSNotificationCenter defaultCenter] addObserver:dropboxManager selector:@selector(dropboxClientUnlinked:) name:FTDidUnlinkAllDropboxClient object:nil];
    });
    return dropboxManager;
}

- (BOOL)isLoggedIn
{
    return [[DBClientsManager authorizedClient] isAuthorized];
}

-(void)authenticateToDropBoxFromController:(UIViewController*)controller
                              onCompletion:(FTDropboxAuthenticateCallback)completionHandler
{
    self.authCallBack = completionHandler;
    [DBClientsManager authorizeFromController:[UIApplication sharedApplication]
                                   controller:controller
                                      openURL:^(NSURL *url) {
                                          [[UIApplication sharedApplication] openURL:url options:@{} completionHandler:nil];
                                      }];
    [[NSNotificationCenter defaultCenter] addObserver:self selector:@selector(didCompleteDropBoxAuthetication:) name:FTDidCompleteDropBoxAuthetication object:nil];
    
    [[NSNotificationCenter defaultCenter] addObserver:self selector:@selector(didCancelDropBoxAuthetication:) name:FTDidCancelDropBoxAuthetication object:nil];
}

-(void)signOutOnCompletionHandler:(GenericSuccessBlock)block
{
    dbSharedSessionUnlink();
    if (block)
    {
        block(YES);
    }
}

-(void)didCancelDropBoxAuthetication:(NSNotification*)notification
{
    [[NSNotificationCenter defaultCenter] removeObserver:self];
    if(self.authCallBack)
    {
        self.authCallBack(NO,YES);
        self.authCallBack = nil;
    }
}

-(void)didCompleteDropBoxAuthetication:(NSNotification*)notification
{
    [[NSNotificationCenter defaultCenter] removeObserver:self];
    if(self.authCallBack)
    {
        self.authCallBack(YES,NO);
        self.authCallBack = nil;
    }
}

-(void)accountInfoOnCompletion:(FTDropboxAccountInfoCallback)completionHandler
{
    if (self.accountInfo && completionHandler)
    {
        completionHandler(self.accountInfo,nil);
    }
    [self.accInfoCallBackArray addObject:completionHandler];
    [self loadAccountInfo];
}


-(void)dropboxClientUnlinked:(NSNotification*)notification
{
    [[NSNotificationCenter defaultCenter] removeObserver:self];
    self.accountInfo = nil;
}

-(void)loadAccountInfo
{
    DBUserClient *client = [DBClientsManager authorizedClient];
    [[client.usersRoutes getCurrentAccount] setResponseBlock:^(DBUSERSFullAccount * _Nullable result, DBNilObject * _Nullable routeError, DBRequestError * _Nullable networkError) {
        if(nil != result) {
            FTBackUpAccountInfo *accountInfo = [[FTBackUpAccountInfo alloc] init];
            accountInfo.name = [[result name] displayName];
            
            [[client.usersRoutes getSpaceUsage] setResponseBlock:^(DBUSERSSpaceUsage * _Nullable result, DBNilObject * _Nullable routeError, DBRequestError * _Nullable networkError) {
                if(nil != result) {
                    switch ([result allocation].tag){
                        case DBUSERSSpaceAllocationIndividual:
                        {
                            accountInfo.consumedBytes = [[result used] longLongValue];
                            accountInfo.totalBytes = [[[[result allocation] individual] allocated] longLongValue];
                        }
                            break;
                        case DBUSERSSpaceAllocationTeam:
                        {
                            accountInfo.consumedBytes = [[[[result allocation] team] used] longLongValue];
                            accountInfo.totalBytes = [[[[result allocation] team] allocated] longLongValue];
                        }
                            break;
                        default:
                            break;
                    }
                }
                
                for (FTDropboxAccountInfoCallback eachInfo in self.accInfoCallBackArray)
                {
                    eachInfo(accountInfo,nil);
                }
                [self.accInfoCallBackArray removeAllObjects];
            }];
            self.accountInfo = accountInfo;
        }
        else {
            for (FTDropboxAccountInfoCallback eachInfo in self.accInfoCallBackArray)
            {
                if (networkError.statusCode.integerValue == 401) {
                    [[NSNotificationCenter defaultCenter] postNotificationName:@"DBLoggedOut" object:nil];
                }
                else {
                    eachInfo(nil,[networkError nsError]);
                }
            }
        }
    }];
}

@end

File: Classes/Mac App Alert/Cloud/Dropbox/FTDropboxManager.swift
//
//  FTDropboxManager.swift
//  Noteshelf
//
//  Created by Simhachalam Naidu on 02/09/20.
//  Copyright © 2020 Fluid Touch Pte Ltd. All rights reserved.
//

import UIKit
import SwiftyDropbox

typealias FTDropboxAuthenticateCallback = (Bool, Bool) -> Void
typealias FTDropboxAccountInfoCallback = (FTBackUpAccountInfo?, Error?) -> Void

@objc class FTDropboxManager: NSObject {
    private var authCallBack: FTDropboxAuthenticateCallback?
    private var accountInfo: FTBackUpAccountInfo?
    private var accInfoCallBackArray: [FTDropboxAccountInfoCallback]?
    
    @objc static let sharedDropboxManager = FTDropboxManager()
   
    override init() {
        super.init()
        self.accInfoCallBackArray = [FTDropboxAccountInfoCallback]()
        NotificationCenter.default.addObserver(self, selector: #selector(dropboxClientUnlinked(_:)), name: NSNotification.Name(rawValue: FTDidUnlinkAllDropboxClient), object: nil)
    }
    
    @objc func isLoggedIn() -> Bool {
        return DropboxClientsManager.authorizedClient != nil
    }
    
    func authenticateToDropBox(from controller: UIViewController?,
                               onCompletion completionHandler: @escaping FTDropboxAuthenticateCallback) {
        authCallBack = completionHandler
        //
        DropboxClientsManager.authorizeFromControllerV2(UIApplication.shared,
                                                        controller: controller,
                                                        loadingStatusDelegate: nil,
                                                        openURL: { url in
                                        UIApplication.shared.open(url, options: [:], completionHandler: nil)
        }, scopeRequest: nil) //TODO: (AK) update to V2 Swift API
        NotificationCenter.default.addObserver(self, selector: #selector(didCompleteDropBoxAuthetication(_:)), name: NSNotification.Name(rawValue: FTDidCompleteDropBoxAuthetication), object: nil)
        NotificationCenter.default.addObserver(self, selector: #selector(didCancelDropBoxAuthetication(_:)), name: NSNotification.Name(rawValue: FTDidCancelDropBoxAuthetication), object: nil)
    }

    @objc func didCancelDropBoxAuthetication(_ notification: Notification?) {
        NotificationCenter.default.removeObserver(self)
        if (authCallBack != nil) {
            authCallBack?(false, true)
            authCallBack = nil
        }
    }

    @objc func didCompleteDropBoxAuthetication(_ notification: Notification?) {
        NotificationCenter.default.removeObserver(self)
        if (authCallBack != nil) {
            authCallBack?(true, false)
            authCallBack = nil
        }
    }
    
    func signOut(onCompletionHandler block: GenericSuccessBlock) {
        DropboxClientsManager.unlinkClients()
        NotificationCenter.default.post(name: Notification.Name(FTDidUnlinkAllDropboxClient), object: nil)
        block(true)
    }
    
    func accountInfo(onCompletion completionHandler: @escaping FTDropboxAccountInfoCallback) {
        if (accountInfo != nil){
            completionHandler(accountInfo, nil)
        }
        accInfoCallBackArray?.append(completionHandler)
        loadAccountInfo()
    }
    
    @objc func dropboxClientUnlinked(_ notification: Notification?) {
        NotificationCenter.default.removeObserver(self)
        accountInfo = nil
    }
    
    func loadAccountInfo() {
        let client = DropboxClientsManager.authorizedClient
        client?.users.getCurrentAccount().response(completionHandler: { (account, error) in
            if let result = account {
                let accountInfo = FTBackUpAccountInfo()
                accountInfo.name = result.name.displayName
                client?.users.getSpaceUsage().response(completionHandler: { (resultValue, routeError) in
                    if let result = resultValue {
                        switch result.allocation {
                        case .individual(let space):
                            accountInfo.consumedBytes = result.used
                            accountInfo.totalBytes = space.allocated
                        case .team(let space):
                            accountInfo.consumedBytes = space.used
                            accountInfo.totalBytes = space.allocated
                        default:
                            break
                        }
                    }
                    self.accInfoCallBackArray?.forEach({ (eachInfo) in
                        eachInfo(accountInfo, nil)
                    })
                    self.accInfoCallBackArray?.removeAll()
                })
                self.accountInfo = accountInfo
            }
            self.accInfoCallBackArray?.forEach({ (eachInfo) in
                switch error {
                case .authError:
                    NotificationCenter.default.post(name: NSNotification.Name("DBLoggedOut"), object: nil)
                case .clientError(let err):
                    eachInfo(nil, err)
                default:
                    eachInfo(nil, nil)
                }
            })
        })
    }
}
